File: agent/src/agent.js
import config from './config/config.js';
import { SystemCollector } from './collectors/systemCollector.js';
import { AppCollector } from './collectors/appCollector.js';
import { APICollector } from './collectors/apiCollector.js';
import { WSTransmitter } from './transmitters/wsTransmitter.js';

/**
 * ServWatch Agent
 * Collects system and application metrics and transmits them to the backend server
 */
class ServWatchAgent {
  constructor() {
    this.config = config;
    this.systemCollector = new SystemCollector();
    this.appCollector = new AppCollector();
    this.apiCollector = new APICollector();
    this.transmitter = new WSTransmitter(
      this.config.server.url,
      this.config.agent.id
    );

    this.collectTimer = null;
    this.transmitTimer = null;
    this.running = false;
    this.systemInfo = null;
  }

  /**
   * Start the agent
   */
  async start() {
    if (this.running) {
      console.log('Agent is already running');
      return;
    }

    console.log('Starting ServWatch Agent...');
    console.log(`Agent ID: ${this.config.agent.id}`);
    console.log(`Server: ${this.config.server.url}`);

    // Connect to server
    this.transmitter.connect();

    // Get system info
    this.systemInfo = await this.systemCollector.getSystemInfo();
    console.log('System:', this.systemInfo.hostname || 'Unknown');

    // Start collection timer
    this.collectTimer = setInterval(() => {
      this.collectAndTransmit();
    }, this.config.agent.collectInterval);

    this.running = true;
    console.log('Agent started successfully');
  }

  /**
   * Stop the agent
   */
  stop() {
    if (!this.running) return;

    console.log('Stopping ServWatch Agent...');

    if (this.collectTimer) {
      clearInterval(this.collectTimer);
      this.collectTimer = null;
    }

    if (this.transmitTimer) {
      clearInterval(this.transmitTimer);
      this.transmitTimer = null;
    }

    this.transmitter.disconnect();
    this.running = false;

    console.log('Agent stopped');
  }

  /**
   * Collect and transmit metrics
   */
  async collectAndTransmit() {
    try {
      // Collect system metrics
      const systemMetrics = await this.systemCollector.collectAll();

      // Collect application metrics
      const appMetrics = this.appCollector.getMetrics();

      // Get API stats if available
      const apiMetrics = this.apiCollector.getMetrics();

      // Combine all metrics into plain object (avoid circular references)
      const metrics = JSON.parse(JSON.stringify({
        cpu: systemMetrics?.cpu || { usage: 0, loadAverage: [0, 0, 0], cores: 0, temperature: 0 },
        memory: systemMetrics?.memory || { used: 0, total: 0, percentage: 0, swapTotal: 0, swapUsed: 0, swapPercentage: 0 },
        disk: systemMetrics?.disk || { usage: 0, readRate: 0, writeRate: 0, io: null, drives: [] },
        network: systemMetrics?.network || { rxRate: 0, txRate: 0, interfaces: [], stats: [] },
        gpu: systemMetrics?.gpu || { controllers: [], count: 0, avgUsage: 0, vramPercentage: 0, maxTemperature: 0 },
        temperatures: systemMetrics?.temperatures || { cpu: 0, cores: [], gpu: [], max: 0 },
        processes: systemMetrics?.processes || { total: 0, running: 0, topByCPU: [], topByMemory: [] },
        app: {
          eventLoopDelay: appMetrics?.eventLoop?.delay || 0,
          eventLoopUtilization: appMetrics?.eventLoop?.utilization || 0,
          heapUsed: appMetrics?.heap?.used || 0,
          heapTotal: appMetrics?.heap?.total || 0,
          heapPercentage: appMetrics?.heap?.percentage || 0
        },
        api: {
          avgResponseTime: apiMetrics?.avgResponseTime || 0,
          p95: apiMetrics?.p95 || 0,
          p99: apiMetrics?.p99 || 0,
          errorRate: apiMetrics?.errorRate || 0,
          requestRate: apiMetrics?.requestRate || 0
        },
        agentId: this.config.agent.id,
        hostname: this.systemInfo?.hostname || 'unknown',
        platform: this.systemInfo?.platform || 'unknown',
        uptime: this.systemInfo?.uptime || 0,
        timestamp: Date.now()
      }));

      if (systemMetrics) {
        // Transmit to server
        this.transmitter.transmit(metrics);

        // Log summary (always log for debugging)
        const cpu = metrics.cpu?.usage || 0;
        const mem = metrics.memory?.percentage || 0;
        const gpu = metrics.gpu?.count > 0 ? `${metrics.gpu.count}x ${metrics.gpu.avgUsage.toFixed(1)}%` : 'N/A';
        const temp = metrics.temperatures?.max > 0 ? `${metrics.temperatures.max.toFixed(0)}°C` : 'N/A';
        const processes = metrics.processes?.total || 0;
        const api = metrics.api?.avgResponseTime || 0;
        const eventLoop = metrics.app?.eventLoopDelay || 0;

        console.log(`Metrics transmitted: CPU ${cpu.toFixed(1)}%, ` +
          `Memory ${mem.toFixed(1)}%, ` +
          `GPU ${gpu}, ` +
          `Temp ${temp}, ` +
          `Procs ${processes}, ` +
          `API ${api.toFixed(0)}ms, ` +
          `Event Loop ${eventLoop.toFixed(2)}ms`);
      }
    } catch (error) {
      console.error('Error in collect and transmit:', error);
    }
  }

  /**
   * Format bytes to human readable
   */
  formatBytes(bytes) {
    if (bytes === 0) return '0 B';
    const k = 1024;
    const sizes = ['B', 'KB', 'MB', 'GB'];
    const i = Math.floor(Math.log(bytes) / Math.log(k));
    return (bytes / Math.pow(k, i)).toFixed(1) + ' ' + sizes[i];
  }

  /**
   * Get agent status
   */
  getStatus() {
    return {
      running: this.running,
      agentId: this.config.agent.id,
      connected: this.transmitter.isConnected(),
      bufferSize: this.transmitter.getBufferSize(),
      systemInfo: this.systemInfo
    };
  }
}

// Main execution
const agent = new ServWatchAgent();

// Handle shutdown
process.on('SIGINT', () => {
  console.log('\\nReceived SIGINT, shutting down gracefully...');
  agent.stop();
  process.exit(0);
});

process.on('SIGTERM', () => {
  console.log('\\nReceived SIGTERM, shutting down gracefully...');
  agent.stop();
  process.exit(0);
});

// Start agent
agent.start().catch(error => {
  console.error('Failed to start agent:', error);
  process.exit(1);
});

export default agent;

File: python-agent/servwatch_agent/agent.py
"""
ServWatch Python Agent
Main agent class for collecting and transmitting system metrics
"""

import asyncio
import logging
import signal
import sys
import time
from typing import Optional

from servwatch_agent.config import get_config
from servwatch_agent.collectors.system import SystemCollector
from servwatch_agent.transmitters.websocket import WSTransmitter

logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


class Agent:
    """Main agent class that orchestrates metric collection and transmission"""

    def __init__(self, config_path: Optional[str] = None):
        """
        Initialize the agent.

        Args:
            config_path: Optional path to configuration file
        """
        self.config = get_config(config_path)
        self.collector = None
        self.transmitter = None
        self.running = False
        self._system_info = None

        # Setup logging
        log_level = self.config.get('logging', 'level', default='INFO')
        logging.getLogger().setLevel(getattr(logging, log_level, logging.INFO))

    def start(self):
        """Start the agent"""
        logger.info("Starting ServWatch Python Agent")
        logger.info(f"Agent ID: {self.config.get('agent', 'id')}")
        logger.info(f"Server: {self.config.get('server', 'url')}")

        # Initialize collector
        enable_gpu = self.config.get('agent', 'enableGPU', default=True)
        self.collector = SystemCollector(enable_gpu=enable_gpu)

        # Get system info once
        self._system_info = self.collector.get_system_info()
        logger.info(f"Hostname: {self._system_info.get('hostname', 'Unknown')}")
        logger.info(f"Platform: {self._system_info.get('platform', 'Unknown')}")
        logger.info(f"CPU Cores: {self._system_info.get('cpu', {}).get('cores', 'Unknown')}")

        # Initialize transmitter
        server_url = self.config.get('server', 'url')
        agent_id = self.config.get('agent', 'id')
        self.transmitter = WSTransmitter(server_url, agent_id)

        # Connect to server
        self.transmitter.connect()

        # Register signal handlers
        signal.signal(signal.SIGINT, self._signal_handler)
        signal.signal(signal.SIGTERM, self._signal_handler)

        # Start collection loop
        self.running = True
        self._run()

    def _run(self):
        """Main collection loop"""
        collect_interval = self.config.get('agent', 'collectInterval', default=1000) / 1000
        transmit_interval = self.config.get('agent', 'transmitInterval', default=1000) / 1000

        last_transmit = time.time()
        last_metrics = None

        while self.running:
            try:
                # Collect metrics
                metrics = self.collector.collect_all()

                if metrics:
                    # Add system info to first transmission
                    if last_metrics is None and self._system_info:
                        metrics['systemInfo'] = self._system_info

                    last_metrics = metrics

                    # Transmit based on interval
                    now = time.time()
                    if now - last_transmit >= transmit_interval:
                        self.transmitter.transmit(metrics)
                        last_transmit = now

                        # Log summary
                        cpu = metrics.get('cpu', {}).get('usage', 0)
                        mem = metrics.get('memory', {}).get('percentage', 0)
                        gpu_count = metrics.get('gpu', {}).get('count', 0)
                        logger.debug(f"Metrics - CPU: {cpu:.1f}%, Memory: {mem:.1f}%, GPUs: {gpu_count}")

                # Sleep for collect interval
                time.sleep(collect_interval)

            except Exception as e:
                logger.error(f"Error in collection loop: {e}")
                time.sleep(collect_interval)

    def _signal_handler(self, signum, frame):
        """Handle shutdown signals"""
        logger.info("Received shutdown signal")
        self.stop()

    def stop(self):
        """Stop the agent"""
        if not self.running:
            return

        logger.info("Stopping ServWatch Python Agent")
        self.running = False

        if self.transmitter:
            self.transmitter.disconnect()

        logger.info("Agent stopped")
        sys.exit(0)


def main():
    """Main entry point"""
    import argparse

    parser = argparse.ArgumentParser(description='ServWatch Python Agent')
    parser.add_argument(
        '--config',
        '-c',
        help='Path to configuration file',
        default=None
    )
    parser.add_argument(
        '--server',
        '-s',
        help='Server URL (overrides config)',
        default=None
    )
    parser.add_argument(
        '--agent-id',
        '-i',
        help='Agent ID (overrides config)',
        default=None
    )
    parser.add_argument(
        '--no-gpu',
        action='store_true',
        help='Disable GPU monitoring'
    )

    args = parser.parse_args()

    # Create agent
    agent = Agent(config_path=args.config)

    # Override with command line args
    if args.server:
        agent.config.config['server']['url'] = args.server

    if args.agent_id:
        agent.config.config['agent']['id'] = args.agent_id

    if args.no_gpu:
        agent.config.config['agent']['enableGPU'] = False

    # Start agent
    try:
        agent.start()
    except KeyboardInterrupt:
        agent.stop()


if __name__ == '__main__':
    main()
